File: src/Torch/ce_label_int.py
import torch
import torch.nn.functional as F


def print_case(title, logits_data, target_data, axis):
    logits = torch.tensor(logits_data, dtype=torch.float64, requires_grad=True)
    target = torch.tensor(target_data, dtype=torch.long)

    logits_for_ce = logits if axis == 1 else logits.movedim(axis, 1)
    loss = F.cross_entropy(logits_for_ce, target, reduction="none")

    loss.sum().backward()

    print(f"\n=== {title} ===")
    print("Logits:")
    print(logits.detach())
    print("Target label int:")
    print(target)
    print("Loss:")
    print(loss.detach())
    print("Logits grad:")
    print(logits.grad)


print_case(
    "3D last axis",
    [
        [[1.0, 2.0, 3.0], [0.0, 1.0, 0.0]],
        [[-1.0, 0.0, 1.0], [3.0, 1.0, -2.0]],
    ],
    [
        [2, 1],
        [0, 0],
    ],
    axis=-1,
)

print_case(
    "3D generic axis 1",
    [
        [[1.0, 2.0], [3.0, 4.0], [5.0, 6.0]],
        [[0.0, 1.0], [2.0, 3.0], [4.0, 5.0]],
    ],
    [
        [2, 0],
        [1, 2],
    ],
    axis=1,
)

File: src/CPP/ce_label_int.php
<?php

use PHP2xAI\Tensor\Tensor;
use PHP2xAI\Runtime\CPP\GraphRuntimeCpp;

include("../../vendor/autoload.php");

function runCeLogitsLabelInt(array $logitsData, array $targetData, int $axis, string $title): void
{
	$y = Tensor::createFromData($logitsData, "y");
	$target = Tensor::createFromData($targetData, "target");

	echo "\n=== {$title} ===\n";
	echo "Logits:\n";
	$y->printData();

	echo "Target label int:\n";
	$target->printData();

	$c = $y->CELogitsLabelInt($target, $axis);

	$graphRuntime = GraphRuntimeCpp::createFromOutputTensor($c);
	$graphRuntime->forward();
	$graphRuntime->backward();
	$graphRuntime->refreshTensorsData();

	echo "Loss:\n";
	$c->printData();

	echo "Target grad:\n";
	$target->printGrad();

	echo "Logits grad:\n";
	$y->printGrad();
}

// Fast path: logits shape [batch, steps, classes], CE along the last axis.
runCeLogitsLabelInt(
	[
		[[1.0, 2.0, 3.0], [0.0, 1.0, 0.0]],
		[[-1.0, 0.0, 1.0], [3.0, 1.0, -2.0]],
	],
	[
		[2, 1],
		[0, 0],
	],
	-1,
	"3D last axis"
);

// Generic axis: logits shape [batch, classes, features], CE along axis 1.
runCeLogitsLabelInt(
	[
		[[1.0, 2.0], [3.0, 4.0], [5.0, 6.0]],
		[[0.0, 1.0], [2.0, 3.0], [4.0, 5.0]],
	],
	[
		[2, 0],
		[1, 2],
	],
	1,
	"3D generic axis 1"
);
